internal/printer/print-to-json.go

```diff
@@ -0,0 +1,241 @@
+	}
+	if n.Name != "" {
+		str += fmt.Sprintf(`,"name":"%s"`, escapeForJSON(n.Name))
+	}
+	if n.Value != "" || n.Type == "attribute" {
+		str += fmt.Sprintf(`,"value":"%s"`, escapeForJSON(n.Value))
+	}
+	if len(n.Attributes) > 0 {
+		str += `,"attributes":[`
+		for i, attr := range n.Attributes {
+			str += attr.String()
+			if i < len(n.Attributes)-1 {
+				str += ","
+			}
+		}
+		str += `]`
+	}
+	if len(n.Directives) > 0 {
+		str += `,"directives":[`
+		for i, attr := range n.Directives {
+			str += attr.String()
+			if i < len(n.Directives)-1 {
+				str += ","
+			}
+		}
+		str += `]`
+	}
+	if len(n.Children) > 0 {
+		str += `,"children":[`
+		for i, node := range n.Children {
+			str += node.String()
+			if i < len(n.Children)-1 {
+				str += ","
+			}
+		}
+		str += `]`
+	}
+	if n.Position.Start.Line != 0 {
+		str += `,"position":{`
+		str += fmt.Sprintf(`"start":{"line":%d,"column":%d,"offset":%d}`, n.Position.Start.Line, n.Position.Start.Column, n.Position.Start.Offset)
+		if n.Position.End.Line != 0 {
+			str += fmt.Sprintf(`,"end":{"line":%d,"column":%d,"offset":%d}`, n.Position.End.Line, n.Position.End.Column, n.Position.End.Offset)
+		}
+		str += "}"
+	}
+	str += "}"
+	return str
+}
+
+func PrintToJSON(sourcetext string, n *Node, opts t.ParseOptions) PrintResult {
+	p := &printer{
+		builder: sourcemap.MakeChunkBuilder(nil, sourcemap.GenerateLineOffsetTables(sourcetext, len(strings.Split(sourcetext, "\n")))),
+	}
+	root := ASTNode{}
+	renderNode(p, &root, n, opts)
+	doc := root.Children[0]
+	return PrintResult{
+		Output: []byte(doc.String()),
+	}
+}
+
+func locToPoint(p *printer, loc loc.Loc) ASTPoint {
+	offset := loc.Start
+	info := p.builder.GetLineAndColumnForLocation(loc)
+	line := info[0]
+	column := info[1]
+
+	return ASTPoint{
+		Line:   line,
+		Column: column,
+		Offset: offset,
+	}
+}
+
+func positionAt(p *printer, n *Node, opts t.ParseOptions) ASTPosition {
+	if !opts.Position {
+		return ASTPosition{}
+	}
+
+	if len(n.Loc) == 1 {
+		s := n.Loc[0]
+		start := locToPoint(p, s)
+
+		return ASTPosition{
+			Start: start,
+		}
+	}
+
+	if len(n.Loc) == 2 {
+		s := n.Loc[0]
+		e := n.Loc[1]
+		start := locToPoint(p, s)
+		end := locToPoint(p, e)
+
+		return ASTPosition{
+			Start: start,
+			End:   end,
+		}
+	}
+	return ASTPosition{}
+}
+
+func attrPositionAt(p *printer, n *Attribute, opts t.ParseOptions) ASTPosition {
+	if !opts.Position {
+		return ASTPosition{}
+	}
+
+	k := n.KeyLoc
+	start := locToPoint(p, k)
+
+	return ASTPosition{
+		Start: start,
+	}
+}
+
+func renderNode(p *printer, parent *ASTNode, n *Node, opts t.ParseOptions) {
+	isImplicit := false
+	for _, a := range n.Attr {
+		if transform.IsImplictNodeMarker(a) {
+			isImplicit = true
+			break
+		}
+	}
+	hasChildren := n.FirstChild != nil
+	if isImplicit {
+		if hasChildren {
+			for c := n.FirstChild; c != nil; c = c.NextSibling {
+				renderNode(p, parent, c, opts)
+			}
+		}
+		return
+	}
+	var node ASTNode
+
+	node.Position = positionAt(p, n, opts)
```

Comment:
Position data is not _super_ robust at the moment and might be wrong in some places. This should work but we'll need a cleanup pass to normalize this data in the parser.